### ISSUE TITLE
Referring to a `Union` type declared in a separate file who's both types are Unknown results in `Illegal type annotation: variable not allowed unless it is a type alias`.

### ISSUE THREAD

**Issue by rowillia** [microsoft/pyright#5621]
**Describe the bug**
Referring to a `Union` type declared in a separate file who's both types are Unknown results in `Illegal type annotation: variable not allowed unless it is a type alias`.

Note this is a simple repro - the real world example is from https://github.com/ray-project/ray/blob/master/python/ray/data/block.py#L58

**Code or Screenshots**
If possible, provide a minimal, self-contained code sample (surrounded by triple back ticks) to demonstrate the issue. The code should define or import all referenced symbols.

`other_types.py`:
```python
from typing import TYPE_CHECKING, TypeAlias, Union

if TYPE_CHECKING:
    import io
    import pathlib

    import pandas
    import pyarrow


KnownTypeUnion = Union["io.BytesIO", "pathlib.Path"]
PartiallyUnknownTypeUnion = Union["io.BytesIO", "pyarrow.Table"]
UnknownTypeUnion = Union["pandas.DataFrame", "pyarrow.Table"]
UnknownTypeTypeAlias: TypeAlias = Union["pandas.DataFrame", "pyarrow.Table"]
```

`test_types.py`:
```python
from typing import TYPE_CHECKING, Union

from .other_types import (
    KnownTypeUnion,
    PartiallyUnknownTypeUnion,
    UnknownTypeTypeAlias,
    UnknownTypeUnion,
)

if TYPE_CHECKING:
    import pandas
    import pyarrow

LocalUnknownTypeUnion = Union["pandas.DataFrame", "pyarrow.Table"]

reveal_type(KnownTypeUnion)
reveal_type(UnknownTypeUnion)
reveal_type(UnknownTypeTypeAlias)
reveal_type(LocalUnknownTypeUnion)
reveal_type(PartiallyUnknownTypeUnion)


def use_known_union(foo: KnownTypeUnion | None):
    pass


def use_unknown_union(foo: UnknownTypeUnion | None):
    pass


def use_unknown_typealias(foo: UnknownTypeTypeAlias | None):
    pass


def use_local_unknown_typealias(foo: LocalUnknownTypeUnion | None):
    pass


def use_partially_unknown_union(foo: PartiallyUnknownTypeUnion | None):
    pass
```

Result:
```
> pyright test_types.py other_types.py
pyright test_types.py other_types.py 
WARNING: there is a new pyright version available (v1.1.303 -> v1.1.318).
Please install the new version or set PYRIGHT_PYTHON_FORCE_VERSION to `latest`

No configuration file found.
No pyproject.toml file found.
Assuming Python platform Darwin
Searching for source files
Found 2 source files
pyright 1.1.303
/Users/roy/test_types.py
  /Users/roy/test_types.py:16:13 - information: Type of "KnownTypeUnion" is "Type[BytesIO] | Type[Path]"
  /Users/roy/test_types.py:17:13 - information: Type of "UnknownTypeUnion" is "Unknown"
  /Users/roy/test_types.py:18:13 - information: Type of "UnknownTypeTypeAlias" is "Unknown"
  /Users/roy/test_types.py:19:13 - information: Type of "LocalUnknownTypeUnion" is "Unknown"
  /Users/roy/test_types.py:20:13 - information: Type of "PartiallyUnknownTypeUnion" is "Type[BytesIO] | Unknown"
  /Users/roy/test_types.py:27:28 - error: Illegal type annotation: variable not allowed unless it is a type alias (reportGeneralTypeIssues)
1 error, 0 warnings, 5 informations 
Completed in 0.485sec
```

If your code relies on symbols that are imported from a third-party library, include the associated import statements and specify which versions of those libraries you have installed.
> pandas==2.0.0
> pyarrow==11.0.0

**VS Code extension or command-line**
Are you running pyright as a VS Code extension, a language server in another editor, integrated into Pylance, or the command-line tool? Which version?

> pyright command line.

**Comment by erictraut** [comment]
This is a consequence of the fact that you're using an old-style type alias which is syntactically identical to a variable assignment. Type checkers like pyright need to use various heuristics to distinguish between the two cases. These heuristics, like most, are fallible. I don't consider this a bug in pyright. 

The workaround is to make sure the types used in the RHS expression are defined or use the more modern `TypeAlias` declaration (introduced in [PEP 613](https://peps.python.org/pep-0613/)) to make it explicit to the type checker that your intent is to define a type alias.
